core/src/main/java/com/linecorp/armeria/common/stream/ByteStreamMessage.java

```diff
@@ -0,0 +1,147 @@
+    }
+
+    /**
+     * Writes this {@link ByteStreamMessage} to the given {@link Path} with {@link OpenOption}s.
+     * If the {@link OpenOption} is not specified, defaults to {@link StandardOpenOption#CREATE},
+     * {@link StandardOpenOption#TRUNCATE_EXISTING} and {@link StandardOpenOption#WRITE}.
+     *
+     * <p>Example:<pre>{@code
+     * Path destination = Paths.get("foo.bin");
+     * ByteStreamMessage streamMessage = ByteStreamMessage.of(...);
+     * streamMessage.writeTo(destination);
+     * }</pre>
+     *
+     * @see StreamMessages#writeTo(StreamMessage, Path, OpenOption...)
+     */
+    default CompletableFuture<Void> writeTo(Path destination, OpenOption... options) {
+        return StreamMessage.super.writeTo(Function.identity(), destination, options);
+    }
+
+    /**
+     * Sets the specified {@link ByteBufAllocator} to this {@link ByteStreamMessage}.
+     * @return {@code this}
+     */
+    ByteStreamMessage alloc(ByteBufAllocator alloc);
+
+    /**
+     * Skips the specified bytes in this {@link ByteStreamMessage}.
+     * <pre>{@code
+     * StreamMessage<HttpData> source = StreamMessage.of(HttpData.ofUtf8("12345"),
+     *                                                   HttpData.ofUtf8("67890"));
+     *
+     * List<HttpData> collected = ByteStreamMessage.of(source).skipBytes(4).collect().join();
+     * assert collected.equals(List.of(HttpData.ofUtf8("5"), HttpData.ofUtf8("67890"));
+     * }</pre>
+     *
+     * @throws IllegalArgumentException if the {@code numBytes} is non-positive.
+     */
+    ByteStreamMessage skipBytes(int numBytes);
+
+    /**
+     * Sets the maximum allowed bytes by this {@link ByteStreamMessage}.
+     *
+     * <pre>{@code
+     * StreamMessage<HttpData> source = StreamMessage.of(HttpData.ofUtf8("12345"),
+     *                                                   HttpData.ofUtf8("67890"));
+     *
+     * List<HttpData> collected = ByteStreamMessage.of(source).takeBytes(6).collect().join();
+     * assert collected.equals(List.of(HttpData.ofUtf8("12345"), HttpData.ofUtf8("6"));
+     * }</pre>
+     *
+     * @throws IllegalArgumentException if the {@code numBytes} is non-positive.
+     */
+    ByteStreamMessage takeBytes(int numBytes);
+
+    /**
+     * Sets the maximum allowed bytes for each {@link HttpData} in this {@link ByteStreamMessage}.
+     *
+     * <pre>{@code
+     * StreamMessage<HttpData> source = StreamMessage.of(HttpData.ofUtf8("12345"),
+     *                                                   HttpData.ofUtf8("67890"));
+     *
+     * List<HttpData> collected = ByteStreamMessage.of(source).bufferSize(3).collect().join();
+     * assert collected.equals(List.of(HttpData.ofUtf8("123"), HttpData.ofUtf8("456"),
+     *                                 HttpData.ofUtf8("789"), HttpData.ofUtf8("0")));
+     * }</pre>
+     *
+     * @throws IllegalArgumentException if the {@code numBytes} is non-positive.
+     */
+    ByteStreamMessage bufferSize(int numBytes);
```

Comment:
I'm somewhat unsure if we should have mutable setters (`alloc()` and `bufferSize()`) rather than overloading the methods. Thoughts?

Reply:
[`12b128f` (#4058)](https://github.com/line/armeria/pull/4058/commits/12b128f212b28f77ab224a5503bc84604337ce1f)
`PathStreamMessageBuilder` was added and only `range()` operation remained in `ByteStreamMessage`.